### PR TITLE
Tweak ImmutableDictionary.TryGetValue to improve throughput

### PR DESCRIPTION
**Pull request by stephentoub**
A few tweaks to ImmutableDictionary to reduce the costs of lookups:
- Changed SortedInt32KeyNode to no longer be recursive, and then inlined it into its two call sites to avoid additional indirection and duplicative branches
- Avoid wrapping and unwrapping key value pairs when comparing the first item in a HashBucket

cc: @AArnott, @agocke 

Benchmark on my machine...

| Name              | Before (us) | After (us) | Improvement | 
|-------------------|-------------|------------|-------------| 
|        Lookup1Int | 175         | 117.7      | 32.74%      | 
|      Lookup100Int | 214.3       | 172.7      | 19.41%      | 
|    Lookup10000Int | 525.6       | 507.7      | 3.41%       | 
|     Lookup1String | 563.6       | 441.9      | 21.59%      | 
|   Lookup100String | 632.3       | 530.7      | 16.07%      | 
| Lookup10000String | 1,477.10  | 1,318.00 | 10.77%      | 

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Attributes.Jobs;
using BenchmarkDotNet.Running;
using System;
using System.Collections.Immutable;
using System.Runtime.CompilerServices;

[MemoryDiagnoser]
[InProcess]
public class Benchmark
{
    private static void Main() => BenchmarkRunner.Run<Benchmark>();

    private ImmutableDictionary<int, int> _oneInt;
    private ImmutableDictionary<int, int> _hundredInts;
    private ImmutableDictionary<int, int> _tenThousandInts;

    private ImmutableDictionary<string, string> _oneString;
    private ImmutableDictionary<string, string> _hundredStrings;
    private ImmutableDictionary<string, string> _tenThousandStrings;

    private string[] _strings;

    [GlobalSetup]
    public void Setup()
    {
        _strings = new string[10_000];
        for (int i = 0; i < _strings.Length; i++) _strings[i] = i.ToString();

        _oneInt = ImmutableDictionary.Create<int, int>().Add(0, 0);
        _oneString = ImmutableDictionary.Create<string, string>().Add(_strings[0], _strings[0]);

        _hundredInts = ImmutableDictionary.Create<int, int>();
        _hundredStrings = ImmutableDictionary.Create<string, string>();
        for (int i = 0; i < 100; i++)
        {
            _hundredInts = _hundredInts.Add(i, i);
            _hundredStrings = _hundredStrings.Add(_strings[i], _strings[i]);
        }

        _tenThousandInts = ImmutableDictionary.Create<int, int>();
        _tenThousandStrings = ImmutableDictionary.Create<string, string>();
        for (int i = 0; i < 10000; i++)
        {
            _tenThousandInts = _tenThousandInts.Add(i, i);
            _tenThousandStrings = _tenThousandStrings.Add(_strings[i], _strings[i]);
        }
    }

    [Benchmark]
    public int Lookup1Int()
    {
        int count = 0;
        for (int i = 0; i < 10000; i++)
        {
            for (int j = 0; j < 1; j++)
            {
                if (_oneInt.TryGetValue(j, out _))
                {
                    count++;
                }
            }
        }
        return count;
    }

    [Benchmark]
    public int Lookup100Int()
    {
        int count = 0;
        {
            for (int i = 0; i < 100; i++)
            {
                for (int j = 0; j < 100; j++)
                {
                    if (_hundredInts.TryGetValue(j, out _))
                    {
                        count++;
                    }
                }
            }
        }
        return count;
    }

    [Benchmark]
    public int Lookup10000Int()
    {
        int count = 0;
        for (int i = 0; i < 1; i++)
        {
            for (int j = 0; j < 10000; j++)
            {
                if (_tenThousandInts.TryGetValue(j, out _))
                {
                    count++;
                }
            }
        }
        return count;
    }

    [Benchmark]
    public int Lookup1String()
    {
        string[] strings = _strings;

        int count = 0;
        {
            for (int i = 0; i < 10000; i++)
            {
                for (int j = 0; j < 1; j++)
                {
                    if (_oneString.TryGetValue(strings[j], out _))
                    {
                        count++;
                    }
                }
            }
        }
        return count;
    }

    [Benchmark]
    public int Lookup100String()
    {
        string[] strings = _strings;

        int count = 0;
        for (int i = 0; i < 100; i++)
        {
            for (int j = 0; j < 100; j++)
            {
                if (_hundredStrings.TryGetValue(strings[j], out _))
                {
                    count++;
                }
            }
        }
        return count;
    }

    [Benchmark]
    public int Lookup10000String()
    {
        string[] strings = _strings;

        int count = 0;
        for (int i = 0; i < 1; i++)
        {
            for (int j = 0; j < 10000; j++)
            {
                if (_tenThousandStrings.TryGetValue(strings[j], out _))
                {
                    count++;
                }
            }
        }
        return count;
    }
}
```

### CODE REVIEW

File: src/System.Collections.Immutable/src/System/Collections/Immutable/SortedInt32KeyNode.cs [hunk 1]
```diff
@@ -200,8 +200,28 @@ internal SortedInt32KeyNode<TValue> Remove(int key, out bool mutated)
         [Pure]
         internal TValue GetValueOrDefault(int key)
         {
-            var match = this.Search(key);
-            return match.IsEmpty ? default(TValue) : match._value;
+            SortedInt32KeyNode<TValue> node = this;
+            while (true)
```
Comment (agocke) [inline src/System.Collections.Immutable/src/System/Collections/Immutable/SortedInt32KeyNode.cs]: Is it not worth sharing this and the below implementation?

Comment (stephentoub) [reply src/System.Collections.Immutable/src/System/Collections/Immutable/SortedInt32KeyNode.cs]: We could.  This helps a bit as it avoids an extra call and extra decision based on whether the node is empty, which would already have been made in the shared helper.  And it really only ends up being a few lines of shared code.  But I can change it if you'd prefer...?

**Comment by ahsonkhan** [comment]
> public class Benchmark

Should we add this perf test to https://github.com/dotnet/performance?

cc @jorive, @adamsitnik

File: src/System.Collections.Immutable/src/System/Collections/Immutable/SortedInt32KeyNode.cs (hunk 1 above)
Comment (agocke) [reply src/System.Collections.Immutable/src/System/Collections/Immutable/SortedInt32KeyNode.cs]: Simple enough that I think it's fine. =

**Comment by jorive** [comment]
@ahsonkhan Yes, benchmarks should be added to the performance repository.

**Comment by ahsonkhan** [comment]
> Yes, benchmarks should be added to the performance repository.

Opened an issue: https://github.com/dotnet/performance/issues/371

Thanks for labeling it. Its a good issue for up-for-grabs.

**Comment by nycdotnet** [comment]
Awesome!  Thank you.

**Comment by bp74** [comment]
Thanks for this optimization. Is there a plan when version 1.6 will be released as a stable package to NuGet?

**Pull request merged**